File: research/experiments/sweep/runs/opt-small-kotlin-sonnet45/src/main/kotlin/optimization/ResultTypes.kt
package optimization

/**
 * Shared types and convergence logic used by all optimization algorithms.
 */

/**
 * Options for optimization algorithms.
 */
data class OptimizeOptions(
    val gradTol: Double = 1e-8,
    val stepTol: Double = 1e-8,
    val funcTol: Double = 1e-12,
    val maxIterations: Int = 1000
)

/**
 * Result of an optimization run.
 */
data class OptimizeResult(
    val x: DoubleArray,
    val `fun`: Double,
    val gradient: DoubleArray?,
    val iterations: Int,
    val functionCalls: Int,
    val gradientCalls: Int,
    val converged: Boolean,
    val message: String
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is OptimizeResult) return false

        if (!x.contentEquals(other.x)) return false
        if (`fun` != other.`fun`) return false
        if (gradient != null) {
            if (other.gradient == null) return false
            if (!gradient.contentEquals(other.gradient)) return false
        } else if (other.gradient != null) return false
        if (iterations != other.iterations) return false
        if (functionCalls != other.functionCalls) return false
        if (gradientCalls != other.gradientCalls) return false
        if (converged != other.converged) return false
        if (message != other.message) return false

        return true
    }

    override fun hashCode(): Int {
        var result = x.contentHashCode()
        result = 31 * result + `fun`.hashCode()
        result = 31 * result + (gradient?.contentHashCode() ?: 0)
        result = 31 * result + iterations
        result = 31 * result + functionCalls
        result = 31 * result + gradientCalls
        result = 31 * result + converged.hashCode()
        result = 31 * result + message.hashCode()
        return result
    }
}

/**
 * Convergence reason (tagged union).
 */
sealed class ConvergenceReason {
    object Gradient : ConvergenceReason()
    object Step : ConvergenceReason()
    object Function : ConvergenceReason()
    object MaxIterations : ConvergenceReason()
    object LineSearchFailed : ConvergenceReason()
}

/**
 * Create default options with optional overrides.
 */
fun defaultOptions(overrides: OptimizeOptions.() -> OptimizeOptions = { this }): OptimizeOptions {
    return OptimizeOptions().overrides()
}

/**
 * Check convergence criteria in order: gradient → step → function → maxIterations.
 * Returns the first matched criterion, or null if none are met.
 */
fun checkConvergence(
    gradNorm: Double,
    stepNorm: Double,
    funcChange: Double,
    iteration: Int,
    opts: OptimizeOptions
): ConvergenceReason? {
    if (gradNorm < opts.gradTol) {
        return ConvergenceReason.Gradient
    }
    if (stepNorm < opts.stepTol) {
        return ConvergenceReason.Step
    }
    if (funcChange < opts.funcTol) {
        return ConvergenceReason.Function
    }
    if (iteration >= opts.maxIterations) {
        return ConvergenceReason.MaxIterations
    }
    return null
}

/**
 * Check if a convergence reason represents successful convergence.
 * True for gradient/step/function; false for maxIterations/lineSearchFailed.
 */
fun isConverged(reason: ConvergenceReason): Boolean {
    return when (reason) {
        is ConvergenceReason.Gradient -> true
        is ConvergenceReason.Step -> true
        is ConvergenceReason.Function -> true
        is ConvergenceReason.MaxIterations -> false
        is ConvergenceReason.LineSearchFailed -> false
    }
}

/**
 * Get human-readable convergence message.
 */
fun convergenceMessage(reason: ConvergenceReason): String {
    return when (reason) {
        is ConvergenceReason.Gradient -> "Gradient norm below tolerance"
        is ConvergenceReason.Step -> "Step size below tolerance"
        is ConvergenceReason.Function -> "Function change below tolerance"
        is ConvergenceReason.MaxIterations -> "Maximum iterations reached"
        is ConvergenceReason.LineSearchFailed -> "Line search failed"
    }
}

File: research/experiments/sweep/runs/opt-full-python-sonnet45/src/result_types.py
"""
Shared types and convergence logic used by all optimization algorithms.
"""

from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class OptimizeOptions:
    grad_tol: float = 1e-8
    step_tol: float = 1e-8
    func_tol: float = 1e-12
    max_iterations: int = 1000


@dataclass
class OptimizeResult:
    x: List[float] = field(default_factory=list)
    fun: float = 0.0
    gradient: Optional[List[float]] = None
    iterations: int = 0
    function_calls: int = 0
    gradient_calls: int = 0
    converged: bool = False
    message: str = ""


@dataclass
class ConvergenceReason:
    kind: str  # "gradient", "step", "function", "maxIterations", "lineSearchFailed"


def default_options(**overrides) -> OptimizeOptions:
    """Create defaults with optional overrides."""
    opts = OptimizeOptions()
    for k, v in overrides.items():
        if hasattr(opts, k) and v is not None:
            setattr(opts, k, v)
    return opts


def check_convergence(
    grad_norm: float,
    step_norm: float,
    func_change: float,
    iteration: int,
    opts: OptimizeOptions,
) -> Optional[ConvergenceReason]:
    """Check criteria in order: gradient -> step -> function -> maxIterations."""
    if grad_norm < opts.grad_tol:
        return ConvergenceReason(kind="gradient")
    if step_norm < opts.step_tol:
        return ConvergenceReason(kind="step")
    if func_change < opts.func_tol:
        return ConvergenceReason(kind="function")
    if iteration >= opts.max_iterations:
        return ConvergenceReason(kind="maxIterations")
    return None


def is_converged(reason: ConvergenceReason) -> bool:
    """True for gradient/step/function; false for maxIterations/lineSearchFailed."""
    return reason.kind in ("gradient", "step", "function")


def convergence_message(reason: ConvergenceReason) -> str:
    """Human-readable message."""
    messages = {
        "gradient": "Converged: gradient norm below tolerance",
        "step": "Converged: step size below tolerance",
        "function": "Converged: function change below tolerance",
        "maxIterations": "Stopped: reached maximum iterations",
        "lineSearchFailed": "Stopped: line search failed",
    }
    return messages.get(reason.kind, f"Unknown convergence reason: {reason.kind}")
